### DISCUSSION TITLE
SPI4 on pi4 no workie

### DISCUSSION THREAD

**Discussion by jschoch** [multigcs/LinuxCNC-RIO#29] [General]
Ok, so... for lots of reasons, none of which are good, i setup a SPI connector board from the pi to a tango 9k and foolishly chose SPI4.

Been poking at it in rio.c.  Very confusing how the pin mapping works, but for those interested there is a map here:

https://www.airspayce.com/mikem/bcm2835/group__constants.html#gga63c029bd6500167152db4e57736d0939afd1fa88cacdbd77188a7fe743e0d34fa

if you are using a pi 2 or 3 then there is this

https://elinux.org/RPi_BCM2835_GPIOs

and for pi4 there is this

https://elinux.org/RPi_BCM2711_GPIOs

So, for my pi4 i'm trying 

```
 // Configure SPI4
    bcm2835_spi_setBitOrder(BCM2835_SPI_BIT_ORDER_MSBFIRST);
    bcm2835_spi_setDataMode(BCM2835_SPI_MODE0);
    bcm2835_spi_setClockDivider(SPI_SPEED);
    bcm2835_spi_chipSelect(BCM2835_SPI_CS_NONE);
    // GPIO 6 on PI4
    bcm2835_gpio_set_pud(RPI_BPLUS_GPIO_J8_29, BCM2835_GPIO_PUD_DOWN);  // MOSI
    // GPIO 5 on PI4
    bcm2835_gpio_set_pud(RPI_BPLUS_GPIO_J8_31 , BCM2835_GPIO_PUD_DOWN); // MISO
    // GPIO 25
    bcm2835_gpio_set_pud(RPI_BPLUS_GPIO_J8_22, BCM2835_GPIO_PUD_UP);            // CS0

```

which looks odd but that seems to be the mapping of the pins but i can't really tell and it doesn't work.

**Comment by multigcs** [comment]
Oh, sorry, never test other SPI devices :(

have you test it with the python test tool ?

`python3 Output/TangNano9K/Firmware/qt_spitest.py`

maybe its simpler in python !?!?

in rio.c, you have also to change the cs pin (not set by the spi driver)

```
#ifdef TRANSPORT_SPI
    int i;

    bcm2835_gpio_fsel(RPI_GPIO_P1_26, BCM2835_GPIO_FSEL_OUTP);
    bcm2835_gpio_write(RPI_GPIO_P1_26, LOW);

    for (i = 0; i < SPIBUFSIZE; i++) {
        rxData.rxBuffer[i] = bcm2835_spi_transfer(txData.txBuffer[i]);
    }
    bcm2835_gpio_write(RPI_GPIO_P1_26, HIGH);
#endif
```
maybe this is the problem ?

**Comment by jschoch** [comment]
I made a new board so i'll close this as unsolved
